### PR TITLE
Refine `source-dirs-on-classpath` heuristic

### LINKED ISSUE #346: Tests fail if run from under a target directory
## Expected behavior
As part of rewrite-clj canary testing, I extract a copy of refactor-nrepl source code somewhere beneath rewrite-clj's `target` directory.

When I run 
```
make test
```
I should see all tests pass.

## Actual behavior
I see the following test failures:
```
...

FAIL in (source-files-with-clj-like-extension-test) (core_test.clj:45)
expected: (pos? (count (filter (fn [f] (-> f .getPath (.endsWith extension))) result)))
  actual: (not (pos? 0))

lein test :only refactor-nrepl.core-test/source-files-with-clj-like-extension-test

FAIL in (source-files-with-clj-like-extension-test) (core_test.clj:45)
expected: (pos? (count (filter (fn [f] (-> f .getPath (.endsWith extension))) result)))
  actual: (not (pos? 0))

lein test :only refactor-nrepl.core-test/source-files-with-clj-like-extension-test

FAIL in (source-files-with-clj-like-extension-test) (core_test.clj:45)
expected: (pos? (count (filter (fn [f] (-> f .getPath (.endsWith extension))) result)))
  actual: (not (pos? 0))

...

FAIL in (sorts-by-frequencies) (form-init11474675976321826582.clj:32)
expected: (= (first utils) (quote refactor-nrepl.util))
  actual: (not (= nil refactor-nrepl.util))

lein test :only refactor-nrepl.ns.namespace-aliases-test/libspecs-are-cached

FAIL in (libspecs-are-cached) (form-init11474675976321826582.clj:42)
expected: (thrown-with-msg? Exception #"Expected!" (sut/namespace-aliases false))
  actual: nil

...

Ran 130 tests containing 29599 assertions.
5 failures, 0 errors.
Tests failed.
Error encountered performing task 'test' with profile(s): 'base,system,provided,dev,1.10,config'
Tests failed.
make: *** [test] Error 1
```

If I instead don't have `target` in my path all tests pass.

## Steps to reproduce the problem
```
mkdir -p foobar/target
cd foobar/target
git clone https://github.com/clojure-emacs/refactor-nrepl.git
cd refactor-nrepl
git reset v3.0.0 --hard
make test
```
You will see tests fail, but if you then:
```
cd ../..
mv target ok
cd ok/refactor-nrepl
make test
```
You will see tests pass.

### clj-refactor.el and refactor-nrepl version information
Running tests from v3.0.0 tag clone of this repository.

### CIDER version information
n/a

### Leiningen or Boot version
```
❯ lein --version
Leiningen 2.9.7 on Java 11.0.13 OpenJDK 64-Bit Server VM
```
### Emacs version
n/a

### Operating system
macOS 10.15.7

## Diagnosis
This was working for refactor-nrepl 2.5.1.

I think the issue is likely caused by new classpath exclusion techniques:
https://github.com/clojure-emacs/refactor-nrepl/blob/a3f9de809b1246101786ec6160df3b8c66585f74/src/refactor_nrepl/core.clj#L62-L72

which excludes any path that contains `target` or `resources` (I'm guessing that `.gitlibs` is less important).

I raised this issue in case it might be affecting your broader user base.
If it is only affecting my particular use case, I can find a work-around for rewrite-clj canary testing.

### PR DESCRIPTION
**Pull request by vemv**
Fixes #346, particularly this nuance: https://github.com/clojure-emacs/refactor-nrepl/issues/346#issuecomment-951338268

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,7 @@
 
 * [#344](https://github.com/clojure-emacs/refactor-nrepl/issues/344): make clean-ns's style closer to the [how to ns](https://stuartsierra.com/2016/08/27/how-to-ns) style.
 * [#333](https://github.com/clojure-emacs/refactor-nrepl/issues/333): skip scanning irrelevant directories in more places (as it was already done for various other functionalities). 
+* [#346](https://github.com/clojure-emacs/refactor-nrepl/issues/346): refine the heuristic for ignoring irrelevant dirs.
```
Comment (bbatsov) [inline CHANGELOG.md:9]: Irrelevant to what? :-)

**Comment by bbatsov** [comment]
The changes seem reasonable, although the changelog entry will probably mean little to most people.

**Comment by vemv** [comment]
Thank! Amended w/ feedback

File: src/refactor_nrepl/core.clj [hunk 2]
```diff
@@ -65,18 +65,28 @@
...
+(def ^:private resources-dir-pattern
+  (re-pattern (str "resources" File/separator "?$")))
+
+(def ^:private target-dir-pattern
+  (re-pattern (str "target" File/separator "?$")))
+
+(defn irrelevant-dir?
+  "Does `f` look like a directory without files that should be analyzed/refactored?"
+  [^File f]
+  (let [s (-> f .toString)]
```
Comment (bbatsov) [inline src/refactor_nrepl/core.clj:77]: Why not just `str`? It calls `.toString` anyways.

File: src/refactor_nrepl/core.clj [hunk 3]
```diff
@@ -65,18 +65,28 @@
...
+    (boolean (or (if *skip-resources?*
+                   (re-find resources-dir-pattern s)
+                   false)
+                 (re-find target-dir-pattern s)
+                 (-> s (.contains ".gitlibs"))))))
```
Comment (bbatsov) [inline src/refactor_nrepl/core.clj:82]: I think in Clojure 1.8 there was a matching function in `clojure.string`.

File: src/refactor_nrepl/core.clj [hunk 4]
```diff
@@ -65,18 +65,28 @@
...
+  (re-pattern (str "resources" File/separator "?$")))
+
+(def ^:private target-dir-pattern
+  (re-pattern (str "target" File/separator "?$")))
+
+(defn irrelevant-dir?
+  "Does `f` look like a directory without files that should be analyzed/refactored?"
+  [^File f]
+  (let [s (-> f .toString)]
+    (boolean (or (if *skip-resources?*
```
Comment (bbatsov) [inline src/refactor_nrepl/core.clj:78]: I'd probably skip this conversion to boolean, but I'm fine either way. I like boolean predicates, but I know most of the community doesn't care what exactly gets returned.

File: src/refactor_nrepl/core.clj (hunk 2 above)
Comment (vemv) [reply src/refactor_nrepl/core.clj:77]: b/c anecdotically it's more performant, and more importantly it signals that I want to invoke File#toString, not an arbitrary string representation (like `pr-str`)

File: src/refactor_nrepl/core.clj (hunk 4 above)
Comment (vemv) [reply src/refactor_nrepl/core.clj:78]: Yeah grey zone, for me the decider was leaving an expressive unit test

File: src/refactor_nrepl/core.clj (hunk 3 above)
Comment (vemv) [reply src/refactor_nrepl/core.clj:82]: Can use, not against it but looks like I'm not adding another commit anyway

In general in projects with zero reflection warnings (an increasing reality 😇) one can afford to call these fearlessly and with full tooling support e.g. Compliment

**Review by bbatsov — APPROVED** [review]

**Pull request merged**